File: src/main/java/fs/math/Vector3.java
package fs.math;

import static java.lang.String.format;

public class Vector3 {
  private float x;
  private float y;
  private float z;

  public Vector3(float x, float y, float z) {
    this.x = x;
    this.y = y;
    this.z = z;
  }

  public static Vector3 vec3() {
    return vec3(0, 0, 0);
  }

  public static Vector3 vec3(float x, float y, float z) {
    return new Vector3(x, y, z);
  }

  public static Vector3 vec3(Vector3 v) {
    return new Vector3(v.x, v.y, v.z);
  }

  public static Vector3 vec3(Vector4 v) {
    return vec3(v.x(), v.y(), v.z());
  }

  public float x() {
    return x;
  }

  public float y() {
    return y;
  }

  public float z() {
    return z;
  }

  public Vector3 x(float x) {
    this.x = x;
    return this;
  }

  public Vector3 y(float y) {
    this.y = y;
    return this;
  }

  public Vector3 z(float z) {
    this.z = z;
    return this;
  }

  public Vector3 set(float x, float y, float z) {
    this.x = x;
    this.y = y;
    this.z = z;

    return this;
  }

  public Vector3 rotate(Quaternion4 quat) {

    float x2 = quat.x() * 2f;
    float y2 = quat.y() * 2f;
    float z2 = quat.z() * 2f;
    float xx2 = quat.x() * x2;
    float xy2 = quat.x() * y2;
    float xz2 = quat.x() * z2;
    float xw2 = quat.w() * x2;
    float yy2 = quat.y() * y2;
    float yz2 = quat.y() * z2;
    float yw2 = quat.w() * y2;
    float zz2 = quat.z() * z2;
    float zw2 = quat.w() * z2;

    return set(
        (1f - (yy2 + zz2)) * x + (xy2 - zw2) * y + (xz2 + yw2) * z,
        (xy2 + zw2) * x + (1f - (xx2 + zz2)) * y + (yz2 - xw2) * z,
        (xz2 - yw2) * x + (yz2 + xw2) * y + (1f - (xx2 + yy2)) * z
    );
  }

  public Vector3 multiply(Matrix4 mat4) {
    return set(
        x * mat4.m00() + y * mat4.m01() + z * mat4.m02() + mat4.m03(),
        x * mat4.m10() + y * mat4.m11() + z * mat4.m12() + mat4.m13(),
        x * mat4.m20() + y * mat4.m21() + z * mat4.m22() + mat4.m23()
    );
  }

  public String toString() {
    return format("[%f %f %f]", x, y, z);
  }

  @Override
  public boolean equals(Object o) {
    if (!(o instanceof Vector3)) {
      return false;
    }

    Vector3 v = (Vector3) o;

    return Math.abs(x - v.x) <= 0.00001f &&
        Math.abs(y - v.y) <= 0.00001f &&
        Math.abs(z - v.z) <= 0.00001f;
  }

  public Vector3 subtract(Vector3 v) {
    return set(x - v.x, y - v.y, z - v.z);
  }

  public Vector3 set(Vector3 v) {
    return set(v.x, v.y, v.z);
  }

  public Vector3 normalize() {
    return multiply(1 / magnitude());
  }

  public Vector3 multiply(float c) {
    return set(x * c, y * c, z * c);
  }

  public float magnitude() {
    return (float) Math.sqrt(x * x + y * y + z * z);
  }

  public Vector3 add(Vector3 v) {
    return set(x + v.x, y + v.y, z + v.z);
  }

  public float dot(Vector3 v) {
    return x * v.x + y * v.y + z * v.z;
  }

  public Vector3 negate() {
    return multiply(-1);
  }
}

File: src/main/kotlin/com/fracturedskies/engine/math/Vector3.kt
package com.fracturedskies.engine.math

data class Vector3(val x: Float, val y: Float, val z: Float) {
  companion object {
    val AXIS_X = Vector3(1f, 0f, 0f)
    val AXIS_NEG_X = Vector3(-1f, 0f, 0f)
    val AXIS_Y = Vector3(0f, 1f, 0f)
    val AXIS_NEG_Y = Vector3(0f, -1f, 0f)
    val AXIS_Z = Vector3(0f, 0f, 1f)
    val AXIS_NEG_Z = Vector3(0f, 0f, -1f)
  }

  operator fun times(mat4: Matrix4): Vector3 {
    return Vector3(
      x * mat4.m00 + y * mat4.m01 + z * mat4.m02 + mat4.m03,
      x * mat4.m10 + y * mat4.m11 + z * mat4.m12 + mat4.m13,
      x * mat4.m20 + y * mat4.m21 + z * mat4.m22 + mat4.m23
    )
  }

  operator fun times(q: Quaternion4): Vector3 {
    val x2 = q.x * 2f
    val y2 = q.y * 2f
    val z2 = q.z * 2f
    val xx2 = q.x * x2
    val xy2 = q.x * y2
    val xz2 = q.x * z2
    val xw2 = q.w * x2
    val yy2 = q.y * y2
    val yz2 = q.y * z2
    val yw2 = q.w * y2
    val zz2 = q.z * z2
    val zw2 = q.w * z2

    return Vector3(
      (1f - (yy2 + zz2)) * x + (xy2 - zw2) * y + (xz2 + yw2) * z,
      (xy2 + zw2) * x + (1f - (xx2 + zz2)) * y + (yz2 - xw2) * z,
      (xz2 - yw2) * x + (yz2 + xw2) * y + (1f - (xx2 + yy2)) * z
    )
  }

  override fun equals(other: Any?): Boolean {
    return when (other) {
      is Vector3 -> Math.abs(x - other.x) <= 0.00001f &&
              Math.abs(y - other.y) <= 0.00001f &&
              Math.abs(z - other.z) <= 0.00001f
      else -> false
    }
  }
}